spack install command incorrectly returning 0 when failing

I have a CI pipeline that builds a few packages daily to make sure that they keep building correctly. The CI script includes the following in a function:

```
spack install $PKGNAME
RET=$?
```

The script then reports the value of RET in a log file.

For the past week or so, I've been hit by [this issue](https://github.com/spack/spack/issues/20454), effectively making all my packages fail to build because automake-1.16.3 could not be built. The odd thing, however, is that Spack doesn't return an error code when failing to build a package.

Here is what spack displays (the package I'm trying to build is called mochi-margo):

```
==> Installing automake-1.16.3-2p5gwby4xivgjq2jwv6cnvjs5ou7k2r4
==> No binary for automake-1.16.3-2p5gwby4xivgjq2jwv6cnvjs5ou7k2r4 found: installing from source
==> Using cached archive: /sandbox/jenkins-ci/workspace/Mochi/mochi-all/spack/var/spack/cache/_source-cache/archive/ce/ce010788b51f64511a1e9bb2a1ec626037c6d0e7ede32c1c103611b9d3cba65f.tar.gz
==> automake: Executing phase: 'autoreconf'
==> automake: Executing phase: 'configure'

==> automake: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16'

2 errors found in build log:
     99       GEN      bin/aclocal-1.16
     100      GEN      doc/aclocal-1.16.1
     101      GEN      bin/automake-1.16
     102      GEN      doc/automake-1.16.1
     103    help2man: can't get `--help' info from aclocal-1.16
     104    Try `--no-discard-stderr' if option outputs to stderr
  >> 105    make: *** [doc/aclocal-1.16.1] Error 126
     106    make: *** Waiting for unfinished jobs....
     107    help2man: can't get `--help' info from automake-1.16
     108    Try `--no-discard-stderr' if option outputs to stderr
  >> 109    make: *** [doc/automake-1.16.1] Error 126

See build log for details:
  /sandbox/jenkins-ci/workspace/Mochi/mochi-all/.spack/stage/autotest/spack-stage-automake-1.16.3-2p5gwby4xivgjq2jwv6cnvjs5ou7k2r4/spack-build-out.txt

==> Warning: Skipping build of mochi-margo-0.9-u7muohgv2xx7ibr2udwusxgzyzwhwtj3 since automake-1.16.3-2p5gwby4xivgjq2jwv6cnvjs5ou7k2r4 failed
```

As you can see in the last line, spack skips the installation of the package I asked it to install, and returns 0, making my CI pipeline think that the package was built correctly.

--- a/lib/spack/spack/installer.py
+++ b/lib/spack/spack/installer.py
@@ -1609,15 +1609,22 @@ class PackageInstaller(object):
         self._cleanup_all_tasks()
 
         # Ensure we properly report if one or more explicit specs failed
-        if exists_errors or failed_explicits:
+        # or were not installed when should have been.
+        missing = [request.pkg_id for request in self.build_requests if
+                   request.install_args.get('install_package') and
+                   request.pkg_id not in self.installed]
+        if exists_errors or failed_explicits or missing:
             for pkg_id, err in exists_errors:
                 tty.error('{0}: {1}'.format(pkg_id, err))
 
             for pkg_id, err in failed_explicits:
                 tty.error('{0}: {1}'.format(pkg_id, err))
 
+            for pkg_id in missing:
+                tty.error('{0}: Package was not installed'.format(pkg_id))
+
             raise InstallError('Installation request failed.  Refer to '
-                               'recent errors for specific package(s).')
+                               'reported errors for failing package(s).')
 
 
 def build_process(pkg, kwargs):

--- a/lib/spack/spack/test/cmd/dev_build.py
+++ b/lib/spack/spack/test/cmd/dev_build.py
@@ -89,7 +89,7 @@ def test_dev_build_until_last_phase(tmpdir, mock_packages, install_mockery):
     assert os.path.exists(str(tmpdir))
 
 
-def test_dev_build_before_until(tmpdir, mock_packages, install_mockery):
+def test_dev_build_before_until(tmpdir, mock_packages, install_mockery, capsys):
     spec = spack.spec.Spec('dev-build-test-install@0.0.0 dev_path=%s' % tmpdir)
     spec.concretize()
 
@@ -103,13 +103,18 @@ def test_dev_build_before_until(tmpdir, mock_packages, install_mockery):
 
         bad_phase = 'phase_that_does_not_exist'
         not_allowed = 'is not a valid phase'
-        out = dev_build('-u', bad_phase, 'dev-build-test-install@0.0.0')
+        not_installed = 'was not installed'
+        out = dev_build('-u', bad_phase, 'dev-build-test-install@0.0.0',
+                        fail_on_error=False)
         assert bad_phase in out
         assert not_allowed in out
+        assert not_installed in out
 
-        out = dev_build('-b', bad_phase, 'dev-build-test-install@0.0.0')
+        out = dev_build('-b', bad_phase, 'dev-build-test-install@0.0.0',
+                        fail_on_error=False)
         assert bad_phase in out
         assert not_allowed in out
+        assert not_installed in out
 
 
 def print_spack_cc(*args):

--- a/lib/spack/spack/test/cmd/install.py
+++ b/lib/spack/spack/test/cmd/install.py
@@ -629,10 +629,11 @@ def test_cache_only_fails(tmpdir, mock_fetch, install_mockery, capfd):
     # libelf from cache fails to install, which automatically removes the
     # the libdwarf build task
     with capfd.disabled():
-        out = install('--cache-only', 'libdwarf')
+        out = install('--cache-only', 'libdwarf', fail_on_error=False)
 
     assert 'Failed to install libelf' in out
     assert 'Skipping build of libdwarf' in out
+    assert 'was not installed' in out
 
     # Check that failure prefix locks are still cached
     failure_lock_prefixes = ','.join(spack.store.db._prefix_failures.keys())

--- a/lib/spack/spack/test/installer.py
+++ b/lib/spack/spack/test/installer.py
@@ -879,7 +879,8 @@ def test_install_failed(install_mockery, monkeypatch, capsys):
     # Make sure the package is identified as failed
     monkeypatch.setattr(spack.database.Database, 'prefix_failed', _true)
 
-    installer.install()
+    with pytest.raises(inst.InstallError, match='request failed'):
+        installer.install()
 
     out = str(capsys.readouterr())
     assert installer.build_requests[0].pkg_id in out
@@ -894,7 +895,8 @@ def test_install_failed_not_fast(install_mockery, monkeypatch, capsys):
     # Make sure the package is identified as failed
     monkeypatch.setattr(spack.database.Database, 'prefix_failed', _true)
 
-    installer.install()
+    with pytest.raises(inst.InstallError, match='request failed'):
+        installer.install()
 
     out = str(capsys.readouterr())
     assert 'failed to install' in out
@@ -1046,7 +1048,9 @@ def test_install_lock_failures(install_mockery, monkeypatch, capfd):
     # Ensure don't continually requeue the task
     monkeypatch.setattr(inst.PackageInstaller, '_requeue_task', _requeued)
 
-    installer.install()
+    with pytest.raises(inst.InstallError, match='request failed'):
+        installer.install()
+
     out = capfd.readouterr()[0]
     expected = ['write locked', 'read locked', 'requeued']
     for exp, ln in zip(expected, out.split('\n')):
@@ -1077,7 +1081,9 @@ def test_install_lock_installed_requeue(install_mockery, monkeypatch, capfd):
     # Ensure don't continually requeue the task
     monkeypatch.setattr(inst.PackageInstaller, '_requeue_task', _requeued)
 
-    installer.install()
+    with pytest.raises(inst.InstallError, match='request failed'):
+        installer.install()
+
     assert b_pkg_id not in installer.installed
 
     out = capfd.readouterr()[0]
@@ -1113,7 +1119,9 @@ def test_install_read_locked_requeue(install_mockery, monkeypatch, capfd):
     const_arg = installer_args(['b'], {})
     installer = create_installer(const_arg)
 
-    installer.install()
+    with pytest.raises(inst.InstallError, match='request failed'):
+        installer.install()
+
     assert 'b' not in installer.installed
 
     out = capfd.readouterr()[0]

